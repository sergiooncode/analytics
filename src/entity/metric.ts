import { Field, Int, ObjectType, registerEnumType } from 'type-graphql';
import { Entity, BaseEntity, PrimaryGeneratedColumn, Column } from 'typeorm';

export enum MetricLevel {
  COMPANY = "COMPANY",
  AGENT = "AGENT"
}

registerEnumType(MetricLevel, {
  name: "MetricLevel"
});

@ObjectType()
@Entity("metric")
export class Metric extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Field(() => String, { nullable: true })
  @Column({ nullable: true })
  name: string;

  @Field(() => Int, { nullable: true })
  @Column({ nullable: true })
  parent_metric: number;

  @Field(() => String)
  @Column()
  level: String;
}