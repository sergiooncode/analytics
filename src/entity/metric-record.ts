import { Field, Int, ObjectType } from 'type-graphql';
import { Entity, BaseEntity, PrimaryGeneratedColumn, Column, OneToOne, JoinColumn } from 'typeorm';
import { Metric } from './metric';

@ObjectType()
@Entity("metric_record")
export class MetricRecord extends BaseEntity {
  @PrimaryGeneratedColumn()
  id: number;

  @Field(() => Int)
  @Column()
  value: number;

  @Field()
  @Column({ type: "timestamp" })
  created_at: Date;

  @Field()
  @OneToOne(() => Metric)
  @JoinColumn({name: "metric_id", referencedColumnName: "id"})
  metric: Metric;
}