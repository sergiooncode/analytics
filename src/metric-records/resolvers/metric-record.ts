import { Query, Resolver } from "type-graphql";
import { getRepository } from "typeorm";
import { MetricRecord } from "../../entity/metric-record";

@Resolver()
export class MetricRecordResolver {

  @Query(() => [MetricRecord])
  metricRecords() {
    const metricRecordRepository = getRepository(MetricRecord);
    return metricRecordRepository.find({
      order: {
        created_at: "DESC"
      }
    });
  }
}