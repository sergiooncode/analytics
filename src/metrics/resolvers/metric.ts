import { Query, Resolver } from "type-graphql";
import { getRepository } from "typeorm";
import { Metric } from "../../entity/metric";

@Resolver()
export class MetricResolver {

  @Query(() => [Metric])
  metrics() {
    const metricRepository = getRepository(Metric);
    return Metric.find();
  }
}