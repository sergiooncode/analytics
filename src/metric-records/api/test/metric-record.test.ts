import { ApolloServer } from "apollo-server-lambda";
import { createTestClient } from "apollo-server-testing";
import gql from "graphql-tag";
import { buildSchema } from "type-graphql";
import { Connection, createConnection, getConnection, getRepository } from "typeorm";
import { Metric, MetricLevel } from "../../../entity/metric";
import { MetricRecord } from "../../../entity/metric-record";
import { MetricResolver } from "../../../metrics/resolvers/metric";
import { MetricRecordResolver } from "../../resolvers/metric-record";

beforeEach(() => {
  return createConnection({
    type: "postgres",
    host: "localhost",
    port: 5432,
    username: "analytics",
    password: "password",
    database: "test_analytics",
    entities: [Metric, MetricRecord],
    logger: "advanced-console",
  });
});

afterEach(() => {
  let conn = getConnection();
  return conn.close();
});

test("query metric records", async () => {
  const metric1 = new Metric();
  metric1.id = 1
  metric1.name = "untaken_calls"
  metric1.level = MetricLevel.COMPANY
  await getConnection().manager.save(metric1);

  const metric2 = new Metric();
  metric2.id = 2
  metric2.name = "forwarded_calls"
  metric2.level = MetricLevel.AGENT
  await getConnection().manager.save(metric2);

  const metricRecord = new MetricRecord();
  metricRecord.id = 1
  metricRecord.value = 100
  metricRecord.metric = metric2
  await getConnection().manager.save(metricRecord);

  const schema = await buildSchema({
    resolvers: [MetricResolver, MetricRecordResolver]
  });

  const serverTest = new ApolloServer({ schema });

  const GET_METRIC_RECORDS = gql`
    query {
      metricRecords {
        value
        created_at
      }
    }
  `;

  const { query } = createTestClient(serverTest);
  
  const res = await query({ query: GET_METRIC_RECORDS });
  
  expect(res).toMatchSnapshot();
});