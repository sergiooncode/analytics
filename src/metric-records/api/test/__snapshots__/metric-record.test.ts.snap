// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query metric records 1`] = `
Object {
  "data": Object {
    "metricRecords": Array [
      Object {
        "created_at": "2021-04-26T04:37:24.345Z",
        "value": 100,
      },
    ],
  },
  "errors": undefined,
  "extensions": undefined,
  "http": Object {
    "headers": Headers {
      Symbol(map): Object {},
    },
  },
}
`;
